Fix to SNAP-2620.

kneeraj opened the pull request:
## Changes proposed in this pull request

The fix is to replace the wrapper, which indicates delete done by the same
transaction, by the gfxd tx entry. During rollback however we would have to
reinstate the committed region entry which was there before this transaction.

## Patch testing

Added dunit for the same

## Is precheckin with -Pstore clean?
Running

## ReleaseNotes changes

None

## Other PRs 

No

Commit:
Fix to SNAP-2620.
The fix is to replace the wrapper, which indicates delete done by the same
transaction, by the gfxd tx entry. During rollback however we would have to
reinstate the committed region entry which was there before this transaction.

Commit:
Changed the test name a bit.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/access/index/MemIndexScanController.java:
```diff
@@ -170,12 +170,11 @@ public final void init(GemFireTransaction tran, MemConglomerate conglomerate,
-          stopSearchOperator, ArrayUtils.objectString(qualifier),
-          tran.getActiveTXState());
+          stopSearchOperator, ArrayUtils.objectString(qualifier));
     }
```
why this logging change?

kneeraj replied:
Noted a task to look into this. This seems to be wrong logging for getting. Active tx state is set properly later. But here it gives:
unexpected change in TXState expected: " + myTX
        + ", thread-local: " + TXManagerImpl.getCurrentTXState();

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/access/GfxdTXStateProxy.java:
```diff
@@ -426,6 +426,13 @@ final void cleanupIndexEntryForDestroy(Object container,
+          deleted = SortedMap2IndexInsertOperation.doMe(null, null, indexContainer,
+              indexKey, (RowLocation) wrapper.getUnderlyingRegionEntry(), true,
+              null, false);
+        }
```
Replacing during insert does not work for some reason? This looks like a hackish way.

kneeraj replied:
For delete followed by insert in a txn an error use to come for unique indexes. So for that the change done is to repplace the wrapperRowLocation put by delete with the GfxdTxEntryState. Check SortedMap2IndexInsertOperation changes. So if you ask me this is not hackish. This only indicates that the wrapper which was suppossed to be there is not anymore as for unique indexes this sequence can create that situation. Please note it is not done for normal index type.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/access/operations/SortedMap2IndexInsertOperation.java:
```diff
@@ -192,80 +186,95 @@ else if (oldValue != null) {
+                // doing a replace and then insert in GfxdTXStateProxy.cleanupIndexEntryForDestroy
+                // Check SNAP-2620 in that class.
+                // TXRegionState txrs = tx.readRegion(((GfxdTXEntryState)value).getDataRegion());
+                // txrs.getToBeReinstatedIndexMap().removeKeyPair(value, oldValue);
```
This looks like the proper fix. Is this missing call to addUniqIdxInfosForReplacedEntries?

kneeraj replied:
Yes as commented, I wanted to do like this but this failed due to to some reason. That's why marked a todo. BTW looks like more fix is required. I extended some testing and at least one more failure is what I am seeing.

sumwale reviewed (commented):
Added few comments. Not sure if current way is clean. We should try to fix the proper way by replacing in insert operation.

Commit:
Fixed issues around (delete --> insert --> update ( both such that index
 key column changes and also some other column changes ) -- then rollback and
    delete. Need to add more test cases.

Commit:
Nullpointer check.

swatisawant committed:
Merge branch 'snappy/master' into SNAP-2620

Commit:
Some more fixes and unit tests.

Commit:
minor change

ahshahid commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/access/operations/SortedMap2IndexInsertOperation.java:
```diff
@@ -191,80 +186,99 @@ else if (oldValue != null) {
+                SQLState.GFXD_OPERATION_CONFLICT, ce, ce.getMessage());
+          } else {
+            if (oldValue instanceof WrapperRowLocationForTxn) {
+              GfxdTXEntryState gfxdtxentry = (GfxdTXEntryState)value;
```
In what situation will the gfxdtxentry not contain the committed region entry?

Commit:
Merge branch 'snappy/master' into SNAP-2620

Commit:
Merge branch 'snappy/master' into SNAP-2620

Commit:
Fix schema in ResultSet metadata

- for routed queries, the schema in ResultSetMetadata was hard-coded to "APP"
- also added a SQLWarning field to pass through any warnings from the source
- correct SQLWarning handling in GemFireResultSet and ij display

Commit:
Fixing build.sh and updated shared libraries with RHEL6 builds

Commit:
Fix build with new IDEA releases

- force output directory of IDEA to be same as gradle output so that it can
  pick up generated java/resource files among others
- generated resource files are now copied to classes output directory rather
  than resources since IDEA does not seem to include resources in CLASSPATH
  when its output directory has been customized like above

Commit:
Catch and throw any Errors like OOMEs during region initialization

don't attempt to cleanup region initialization in case of fatal error like OOME
since that itself can get stuck

ahshahid committed:
Checking if the entry causing constraint violation is destroyed or re… (#441)

* Checking if the entry causing constraint violation is destroyed or removed in a separate thread with synch taken on region entry, to avoid dead lock.

Commit:
Increase history file size to 50000 lines

- skip init to set history max-size else it invokes load() in constructor
  that truncates the file to default 500 lines
- update jline to 2.14.6 for this new constructor (https://github.com/jline/jline2/issues/277)

Commit:
Patch from Sumedh for SNAP-2641

lizygeogy committed:
Update README.md

As suggested by Asif:
:dunit replaced to :dunitTest 
:gemfirexd-tools  replaced with :snappy-store:snappydata-store-tools

Commit:
Add CACHE/UNCACHE/RECACHE/RESET for routing

Commit:
minor change to stdout message in startup

Commit:
[SNAP-2657] fix performance problems with statistics SampleCollector (#444)

For cases having large number of regions (like 200 or more), SampleCollector runs in a hot loop
trying to lookup every statistic in every monitor. Latter is created for every metric/gauge in every
statistic instance and thus runs into thousands (e.g. ~10K with 500 tables) so overall lookups
become like 5 million in every run.

While the design of having a monitor for every metric/gauge for every statistic instance, and looking
up against it in case of any updates itself is inefficient, the changes below bring the performance
within acceptable range with tables of the order of 1000 or so which is sufficient for practical purposes.

## Changes proposed in this pull request

- Skip including statistics for checking that have seen no updates. This reduces the number of changes
  from 500 to single digits and improve the performance by two orders or magnitude or more.
  Actual number of statistics with updates in a second are expected to be small on an average.
- avoid lookup in hive metastore for tableType column in SYSTABLES scan and instead use
  PartitionedRegion.needsBatching()
- avoid lookup of hive metastore for isRowTable() since calling points already have the HiveTable
- renamed "needsBatching()" to "isRowBuffer()" and cleaned up checks for column/row buffers to
  consistently use LocalRegion.isInternalColumnTable() and isRowBuffer()
- add methods to return all PartitionedRegions and clear PartitionedRegion.isRowBuffer since it may
  have been cached incorrectly when column store for a table was not yet created
- add routing for DESCRIBE EXTENDED
- reset isRowBuffer flag after column store initialization to set correctly after recovery

ahshahid committed:
changing the TABLE_TYPE data type to varchar(16) (#443)

Commit:
Updates to remove member ThreadLocals

Some ThreadLocals used by code are member variables that hold reference to the object
causing a memory leak. Changes to remove those ThreadLocals causing them to be remove
from ThreadLocalMap instead of just setting to null/invalid when done.

hemanthmeka committed:
Snap 2503 (#442)

* adding jvmkill.c and also merged it into libgemfirexd64.so and libgemfirexd.so

* changes related to review comments and enhances

* replacing realpath command with absPath()

* added generation of ibgemfirexd*.dylib for Mac

* adding gemfirexd(64).dylib files

* recompiled libgemfirexd so files on centos with logging in jvmkill

* minor fixes

* logging heap histogram on critical event

* adding libgemfirexd dylib files

* adding .so files compiled on centos

* incorporating review suggestions

* implementing review suggestions

Commit:
minor formatting changes

Commit:
Cleanup of log-level naming

- consistently allow for both Log4J format and older GemFire format log-level names
- add "snappydata.numServersOnHost" property to scale down default memory-size
  allocated to a server proportionately
- updated some dependencies in build.gradles to avoid unnecessary recompilations

Commit:
Changes to dump file names and creation

- changed name of "jvmkill.log" to include pid: jvmkill_pid<pid>.log
- create jvmkill log on demand and not on load; re-generated shared libraries on CentOS6
- changed name of jmap dump files to match the convention used by -XX:+HeapDumpOnOutOfMemoryError
  and added a "-" to separate the PID from the random ID

Commit:
avoid double hash in the file name

Commit:
* Version 1.6.2.1

Commit:
jvmkill.c changed. recompiling libgemfirexd dylib files

Commit:
SNAP-2654 : Fix for memory leak (#445)

* Fix for memory leak

  Reaper thread now removes any entry for which there
  exists a committed entry (including in RegionMap) with higher version
  required by same set of transactions as by itself

* use date/time in the file name for jmap dumps

Commit:
Cleanup thread-local pooled connection when it is returned to pool

Added a custom JdbcInterceptor to close the thread-local connection
(now moved to TomcatConnectionPool from ClientPoolDriver).
Always ensure that this interceptor is added even if user has specified
a custom list of interceptors.

Commit:
Add handling for missing log4j log-levels

also added SnappyData copyright header to store-log4j.properties

Commit:
Changed the year frim 2017 to 2018 in license headers.

Commit:
Dump UMM stats for GemXD-specific low memory exceptions too

snappy-sachin committed:
Fixing getColumns call for ODBC meta-data GETINFO

- the new SYS.VTIS addition broke the translation of JDBC to ODBC query
- for query routing to SnappyData, passing the exception if any and logging
  at debug-level using log4j (which can be turned on dynamically if required)

snappy-sachin committed:
SNAP-2602 : On snappy UI, add column named "Overflown Size"/ "Disk Size" in Tables. (#440)

* Changes for SNAP-2602:
 - SnappyRegionStats is updated for holding tables size spillover to disk i.e. tables overflown size.
 - SnappyRegionStatsCollectorFunction is updated to set tables size spillover to disk.
 - and relevant code changes.

snappy-sachin committed:
Changes for SNAP-2612: (#431)

- In order to display external tables fully qualified name, adding variables into SnappyExternalTableStats to hold tables schema and fully qualified names.

Commit:
Show all the different table types for the generic "TABLE" type

When the standard table type "TABLE" is given, then show all of
row, column, external, sample, stream and topK tables. This also
fixes the issue of tables not showing up in "show tables" with
snappy-shell when using hive-thriftserver

Commit:
Fixing failures introduced by merges for SNAP-2602 and SNAP-2654

AnalyzeSerializablesJUnitTest failure introduced by merge of PR#440 for SNAP-2602
SnappyRegionStats cannot change older version serializations, rather new ones should be added

SnapshotTransactionTest failure due to incorrect cast of Queue to NonLocalRegionEntry
introduced by merge of PR#445 for SNAP-2654

Commit:
Meta-data statement routing

- always route statements like "show tables/views...", "describe" to lead from snappy-shell
  to give consistent results (between Spark execution and this, as well as extensions like
      "describe extended" which are routed in any case)
- update completion lists in shell for above
- make table type matching in EmbedDatabaseMetaData as case-insensitive

Commit:
[Snap 2654] Remove duplicate entries from oldEntryMap (#446)

* Fix for memory leak in oldEntryMap
  When no transactions are running, take a write lock on rvv and clear
the map
  In some cases, entry in oldEntryMap and in region could have same
entryVersion. In that case, do clear the entry from oldEntryMap.
  In some cases, specially after truncate the entry in region could have
lower entryVersion. In that case, find out if there are any transaction
which requires the entry in oldEntryMap and if not so then delete it
from oldEntryMap
* Return TOMBSTONE when no entry present in oldEntryMap
* If entry is present in region, then check if it is valid for all
running transactions. If it is, then remove the entry present in
oldEntryMap.

* Account for the memory while clearing oldEntryMap

Commit:
Remove the code for clearing the map under lock (#449)

If no tx is running, we used to take a lock and iterate over map to remove each entry.
There was a bug introduced where entry was not removed from queue
  but the memory was being released for it.
No need of this block, as anyway we are iterating over the map which next block of code does.

Commit:
Allow clearing buffers partially (e.g. after an expand) (#447)

- allow clearing buffers partially (e.g. after an expand)
- dump generated code to separate file
- also a minor change to AbstractRegionEntry._setValue to quickly handle the case when
  current value is null/token and incoming value is a SerializedDiskBuffer
- don't take locks in DiskId.toString to allow logging them safely without deadlocks
- fix failure in BugsTest.testPreBlow running on local machines
- update CDT settings

bisenpiyush committed:
[SNAP-2591] Implemented failover in ODBC. (#439)

* * Implemented failover in ODBC.
For this, created new class ControlConnection.cpp equivalent to JDBC side. After this successful implementation, locator port number can be provide in the connection string.

Commit:
Always wrap exceptions from lead since server may not have the required classes

also avoid creating generatedcode.log in clients

Commit:
Store side changes for catalog cleanup (#452)

Store side changes for catalog cleanup. Primarily addition of generic updateCatalogMetadata
and getCatalogMetadata methods to SnappyHiveCatalog (renamed to StoreHiveCatalog).
The StoreHiveCatalog internally now uses the SnappyExternalCatalog instance instead of having
its own hive client. The two methods mentioned now map directly to corresponding methods
in Spark's ExternalCatalog interface for querying and updating the catalog.

The smart connector has now switched entirely to use these two methods (in ConnectorExternalCatalog) 
which no longer uses specific system procedures for different DDLs or its own hive client for querying.
Two procedures for querying and updating the catalog respectively have been added. This allows for
consistent access control checks for all catalog access.

Thrift IDL has been enhanced to add the argument/result types of these two procedures. While the
procedures themselves are just plain JDBC/ODBC ones and do not have specific thrift calls, keeping
the argument/result serialization allows for ease of backward compatibility in future releases.

## Changes proposed in this pull request

- support for all existing SnappyExternalCatalog methods in StoreHiveCatalog
- support for rename table, drop/rename partitions in thrift definition to enable using with
  connector catalog; can be used by Spark's Parquet/ORC tables
- fix GfxdTXStateProxy rollback to send to locators too for hive meta-store tables that are also
  persisted on locators
- replaced koloboke maps with eclipse collections since that project has been dead for a while now;
  it is somewhat slower and bloats product by ~6MB but looks like best choice (fastutil is too large)
- removed long obsolete transaction exception and other classes related to pre GFE 7.0 transactions
- removed skipLocks in ExternalCatalog where it should always be passed as true
- fixed TXBatchMessage handling in index management:
  - these batched inserts are converted to put DMLs, so handle those like puts in GfxdTXEntryState
    instead of failing with AssertionError
- more cleanups and added more information to catalog stale exception
- rollback TX in case of failure in flush buckets in pre-commit
- avoid creating generatedcode.log in clients
- correct length handling after a free in ClientLobs

Added system procedure to create/drop reservoir region
- Now using a persistent system procedure for create/drop of reservoir region.
  This is to allow new or restarted members to create the region. Also added
  schema permission check.
- Fixed a case of unique constraint violation even with skip-constraints=true
- update ws.rs-api dependency
- minor change to avoid possible NPE when node is going down
- minor change to log4j.properties

Commit:
[SNAP-2789] use updated package for StoreHiveCatalog (#453)

- updated package path of StoreHiveCatalog as per changes on snappydata
- use dataSourcePath for SOURCEPATH in SYS.HIVETABLES which contains either the path
  for file-based tables or URL for JDBC

Commit:
Removed trove based "TStateless" classes

Moved usages of TStateless classes over to eclipse-collections.
Also removed bunch of unused trove classes as well as replaced some by eclipse-collections.

Commit:
Removing more trove classes that were not being used

Commit:
Updating gradle to version 5.0

- updated gradle build files to version 5.0
- updated regression scripts for the new compiled classes layout in gradle 5.0
- removed dependency on javax.mail-api jar and commented out the usage in obsolete
  unused AdminDistributedSystemJmxImpl
- removed unused gemfire-web module

Commit:
- Changed the default admin user to gemfire1.

Commit:
- Modified the conf to have the test run with/without security.

Commit:
change javax.servlet-api version to 3.0.1

Commit:
Adding LDAP test for names having escape characters

Commit:
Added gradle property for both architecutre x86 and x64
Added downloaded version 3.4.3 for de.undercouch

Commit:
Refactored the build.gradle

Commit:
Changes to correct backward compatibility for reservoir regions

- skip send/persistence of create/drop reservoir region procedure if the operation is unsuccessful locally
- avoid heap dump on OOM with junit tests that causes many hprofs due to deliberate CRITICAL_UP tests

ahshahid committed:
added schema creation pattern for routing to lead node (#462)

ahshahid committed:
Snap 2760 (#455)

* changes for getting ldap groups for a user. The implementation first checks if "memberOf" attribute is present in the user's information ( as in case of AD Groups). If the information is available that is used to get names of Ldap Groups else the code recurses through the ldap group tree.

dshirish committed:
[SNAP-2956] Added a constructor with cause (#467)

Commit:
Disable load-balance by default on servers (#470)

* Check the ControlConnection and if it on a server then disable load-balance by default (and close
  the ControlConnection)
* added test for load-balance=false
* updated tests to deal with thrift and added checks that default on server
  connections is load-balance=false
* route "DROP DATABASE" DDL

Commit:
fix occasional NPE in ClientService init

Commit:
Change the default auto reconnect to false (#468)

* Change the default auto reconnect to false, for the locator it is still set to true

Commit:
updates to handle capacity vs limit of DirectByteBuffers correctly in case they are different

Commit:
avoid double printing of test start/stop when run from SnappyData

Commit:
Scalability snapshot (#466)

* Changes related to oldEntryMap and snapshot to store NonLocalRegionEntry instead of BlockingQueue

* Changes for fast cleaning of oldEntryMap Also reduces the overhead by first storing just the NonLocalRegionEntry and later a collection

* Change the old entry cleaner interval

ahshahid committed:
possible approach to fix AQP-295 (#469)

Fixing the issue of reservoir region for sample being skipped during ddl replay. Using a postProcess flag to delay the execution of  GfxdSystemProcedureMessages with postProcess flag as true & replaying those messages , once the boot of the data base has happened.

Commit:
Adding backtick also for quoted identifier.
Partial part of PR https://github.com/SnappyDataInc/snappy-store/pull/471

Commit:
fix some dunit test failures

explicitly use DRDA connections in some tests that don't boot local VM so don't
set thrift-default=false in GemXD mode

Commit:
Set auto reconnect to true in test (#472)

As default in the product has changed the test was failing

Commit:
* Version changes

Commit:
[SNAP-2985] Fixes for multiple inconsistency issue in snapshot (#473)

* Were returning RegionEntry to the iterator, which could have its value changed by a different writer before iterator could consume it
* The NonLocalRegionEntry had oldDiskId and ColumFormatValue used to read from DiskEntry instead of the value in memory. Setting that to null now, which means the value in memory will be read
 * Used CustomEntryConcurrentHashMap to make sure we remove entry only if the same entry is present in the map
* Similarly, made sure that before a queue is removed, its size is 0 and anytime an entry is added to queue  it is put again in the map for atomicity
* While checking whether an entry in region exists in the snapshot of a tx, avoided the entry that was being changed so that due to uncommitted entry a valid entry doesn't get removed from the oldEntryMap
* Added tests where bulk puts and bulk reads are checked for atomicity
* Don't account for SerializeDiskBuffer in off-heap
* For ObjectiStore use different getExecRow.

Commit:
Fix reference count when creating NonLocalRegionEntry (#474)

Problem with using newEntryWithoutFaultin when creating NonLocalRegionEntry is that it leads to an extra reference count. But higher layer cannot do extra release in all cases because when entry is read from oldEntriesMap then the extra release will be incorrect. Also the extra reference count cannot be released immediately after this call because if the value has been read from disk without fault-in then it will have only one refCount so will get released immediately before even being read by iterator.

The change to create NLRE in all cases leads to a performance problem when data is partially on disk since it will be read in essentially random order. This will be fixed in a subsequent PR.

## Changes proposed in this pull request

Use NLRE.newEntry instead of newEntryWithoutFaultin to create a snapshot of entry.

Commit:
* Updated some metainfo in prep for 1.1.0 release.

snappy-sachin committed:
disable load-balance by default in native driver

Commit:
* Select product name to be printed before snappy shell starts, depending upon enterprise flag.
* Discussed with @kneeraj

Commit:
Changes for lower-case naming (#477)

- adjust changes that change SnappyData to use lower-case identifiers
- change quoted identifiers to upper-case:
  for consistency, always use upper-case identifier names because Spark side
  will do case-insensitive analysis by default
- also fixed handling of double backtick conversion to single
- add support for OpenJDK in hydra framework
- fix schema name handling in SnappyRegionStatsCollectorFunction

smahajan05 committed:
Snap 2707 (#459)

* Fixes for SNAP-2707

* Added query routing code for put into values of column table

* changes for SNAP-2707
- Modified exception message.
- Added routeQuery parameter while checking for routing a query.

* Minor change

* Added formatting changes

* Added implementation for returning count of put into query on column table
Incorporated review comments

* Incorporated review comments

Commit:
fix occasional failure in GfxdTableMBeanDUnit

Commit:
Add support for OpenJ9

Commit:
minor change to prevent possible NPE

dshirish committed:
[SNAP-3007] Delete without parameters using PreparedStatement fails to remove data (#481)

* Fix for SNAP-3007: Delete without parameters using PreparedStatement fails to remove data
